sdk/core/core-paging/test/getPagedAsyncIterator.spec.ts

```diff
@@ -0,0 +1,129 @@
+      assert.fail(`should not get here but got: ${val}`);
+    }
+  });
+
+  it("should return an iterator over an non-empty collection", async function() {
+    const collection = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+    const iterator = buildIterator(collection);
+    const expected = [];
+    for await (const val of iterator) {
+      expected.push(val);
+    }
+    assert.deepEqual(expected, collection);
+  });
+
+  it("should return an iterator over an non-collection", async function() {
+    const iterator = buildIterator({});
+    for await (const val of iterator) {
+      assert.deepEqual(val, {});
+    }
+  });
+
+  it("should return an iterator over no pages", async function() {
+    const iterator = buildIterator([]);
+    for await (const val of iterator.byPage({ maxPageSize: 5 })) {
+      assert.deepEqual(val, []);
+    }
+  });
+
+  it("should return an iterator over multiple pages (collections)", async function() {
+    const collection = Array.from(Array(10), (_, i) => i + 1);
+    let currIndex = 0;
+    const pagedResult: PagedResult<number[]> = {
+      firstPageLink: "",
+      async getPage(_path, maxPageSize) {
+        const top = maxPageSize || 5;
+        if (currIndex < collection.length) {
+          const res = collection.slice(currIndex, Math.min(currIndex + top, collection.length));
+          const nextPageLink = top < collection.length - currIndex ? "nextPageLink" : undefined;
+          currIndex += top;
+          return Promise.resolve({
+            page: res,
+            nextPageLink: nextPageLink
+          });
+        } else {
+          throw new Error("should not get here");
+        }
+      }
+    };
+    const iterator = getPagedAsyncIterator(pagedResult);
+    let receivedItems = [];
+    for await (const val of iterator) {
+      receivedItems.push(val);
+    }
+    assert.deepEqual(receivedItems, collection);
+    currIndex = 0;
+    let pagesCount = 0;
+    const maxPageSize = 5;
+    const receivedPages: any[] = [];
+    for await (const val of iterator.byPage({ maxPageSize: maxPageSize })) {
+      ++pagesCount;
+      assert.isAtMost(val.length, maxPageSize);
+      receivedPages.push(val);
+    }
+    assert.equal(pagesCount, Math.ceil(collection.length / maxPageSize));
+    assert.deepEqual([].concat(...receivedPages), collection);
+  });
+
+  it("should return an iterator over multiple pages (non-collections)", async function() {
+    const maxPageSize = 5;
+    const collection = Array.from(Array(10), (_, i) => i + 1);
+    let currIndex = 0;
+    const pagedResult: PagedResult<Record<string, unknown>> = {
+      firstPageLink: "",
+      async getPage(_path, maxPageSize) {
+        const top = maxPageSize || 5;
+        if (currIndex < collection.length) {
+          const res = collection.slice(currIndex, Math.min(currIndex + top, collection.length));
+          const nextPageLink = top < collection.length - currIndex ? "nextPageLink" : undefined;
+          currIndex += top;
```

Comment:
normally, iterators wouldn't be able to rely on a closure variable to track the next index, you'd have to encode it into the `nextPageLink`, right? Maybe it's better to do so in this test to validate that such a thing works as expected. Otherwise you could have a bug where the next link isn't sent back properly but this test would still pass, right?

Reply:
Done in https://github.com/Azure/azure-sdk-for-js/pull/16774/commits/3b3c122ac353f40cb2183a8a5af92eb1343fe682